thrust/thrust/detail/random_bijection.h

```diff
@@ -0,0 +1,173 @@
+    for (std::uint32_t i = 0; i < num_rounds; i++)
+    {
+      std::uint32_t hi, lo;
+      constexpr std::uint64_t M0 = UINT64_C(0xD2B74407B1CE6E93);
+      mulhilo(M0, state[0], hi, lo);
+      lo       = (lo << (right_side_bits - left_side_bits)) | state[1] >> left_side_bits;
+      state[0] = ((hi ^ key[i]) ^ state[1]) & left_side_mask;
+      state[1] = lo & right_side_mask;
+    }
+    // Combine the left and right sides together to get result
+    return (static_cast<std::uint64_t>(state[0]) << right_side_bits) | static_cast<std::uint64_t>(state[1]);
+  }
+
+private:
+  // Perform 64 bit multiplication and save result in two 32 bit int
+  static _CCCL_HOST_DEVICE void mulhilo(std::uint64_t a, std::uint64_t b, std::uint32_t& hi, std::uint32_t& lo)
+  {
+    std::uint64_t product = a * b;
+    hi                    = static_cast<std::uint32_t>(product >> 32);
+    lo                    = static_cast<std::uint32_t>(product);
+  }
+
+  // Find the nearest power of two
+  static _CCCL_HOST_DEVICE std::uint64_t get_cipher_bits(std::uint64_t m)
+  {
+    if (m <= 16)
+    {
+      return 4;
+    }
+    std::uint64_t i = 0;
+    m--;
+    while (m != 0)
+    {
+      i++;
+      m >>= 1;
+    }
+    return i;
+  }
+
+  static constexpr std::uint32_t num_rounds = 24;
+  std::uint64_t right_side_bits;
+  std::uint64_t left_side_bits;
+  std::uint64_t right_side_mask;
+  std::uint64_t left_side_mask;
+  std::uint32_t key[num_rounds];
+};
+
+//! \brief Adaptor for a bijection to work with any size problem. It achieves this by iterating the bijection until
+//! the result is less than n. For feistel_bijection, the worst case number of iterations required for one call to
+//! operator() is O(n) with low probability. It has amortised O(1) complexity.
+//! \tparam IndexType The type of the index to shuffle.
+//! \tparam Bijection The bijection to use. A low quality random bijection may lead to poor work balancing between calls
+//! to the operator().
+template <class IndexType, class Bijection = feistel_bijection>
+class random_bijection
+{
+private:
+  static_assert(::cuda::std::is_integral_v<IndexType>, "IndexType must be an integral type");
+  static_assert(::cuda::std::is_convertible_v<IndexType, typename Bijection::index_type>,
+                "IndexType must be convertible to Bijection::index_type");
+
+  Bijection bijection;
+  IndexType n;
+
+public:
+  using index_type = IndexType;
+
+  template <class URBG>
+  _CCCL_HOST_DEVICE random_bijection(IndexType n, URBG&& g)
+      : bijection(n, ::cuda::std::forward<URBG>(g))
+      , n(n)
+  {}
+
+  _CCCL_HOST_DEVICE IndexType operator()(IndexType i) const
+  {
+    auto upcast_i = static_cast<typename Bijection::index_type>(i);
+    auto upcast_n = static_cast<typename Bijection::index_type>(n);
+
+    // If i < n Iterating a bijection like this will always terminate.
+    // If i >= n, then this may loop forever.
+    // TODO Do we need a runtime check apart from this assert?
```

Comment:
I would appreciate some feedback here on the best way to handle this case, I see a hang as one of the worse errors we could cause, but as with all things C++ go past the end at your own risk

Reply:
This is already a fairly expensive operation from calling the bijection, possibly multiple times. I think adding a runtime branch would make sense here. We can also leave the assert to help catch issues, but change it to `_CCCL_ASSERT(cond, msg)`.

```
if (upcast_i >= upcast_n)
{ // Avoid infinite loop.
  _CCCL_ASSERT(false, "index out of range");
  return upcast_i;
}

// Rest of impl
```

This will at least return a unique value per input index, even if not shuffled.